src/utils/daos_hdlr.c

```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
-			       &src_file_dfs->dfs);
-		if (rc) {
-			fprintf(stderr, "dfs mount on source failed: %d\n", rc);
-			D_GOTO(err_src_dfs, rc);
+		if (is_posix_copy) {
+			rc = dfs_mount(ca->src_poh, ca->src_coh, O_RDWR,
+				       &src_file_dfs->dfs);
+			if (rc) {
+				fprintf(stderr, "dfs mount on source "
+					"failed: %d\n", rc);
+				D_GOTO(err_src, rc);
+			}
 		}
 	}
 
+	/* get source container type */
+	rc = dm_get_cont_prop(ca->src_coh, sysname, src_cont_info,
+			      ca->cont_prop_layout, &ca->cont_layout);
+	if (rc != 0) {
+		fprintf(stderr, "failed to query source "
+			"container: %d\n", rc);
+	}
+
+	/* get allocated oid on source container to set on destination */
+	rc = dm_get_cont_prop(ca->src_coh, sysname, src_cont_info,
+			      ca->cont_prop_oid, &ca->cont_oid);
+	if (rc != 0) {
+		fprintf(stderr, "failed to query source "
+			"container: %d\n", rc);
+	}
+
+	props = daos_prop_alloc(2);
+	if (props == NULL) {
+		fprintf(stderr, "Failed to allocate prop (%d)", rc);
+		D_GOTO(out, rc);
+	}
+	props->dpp_entries[0].dpe_type = ca->cont_prop_layout;
+	props->dpp_entries[0].dpe_val = ca->cont_layout;
+
+	/* only set max oid on created dest containers
+	 * if this is a non-posix copy
+	 */
+	if (!is_posix_copy) {
+		props->dpp_entries[1].dpe_type = ca->cont_prop_oid;
+		props->dpp_entries[1].dpe_val = ca->cont_oid;
+	}
+
 	/* open dst pool, dst cont, and mount dfs */
 	if (dst_file_dfs->type == DAOS) {
-		rc = daos_pool_connect(fa->dst_p_uuid, sysname,
-				       DAOS_PC_RW, &fa->dst_poh, NULL, NULL);
-		if (rc != 0) {
-			fprintf(stderr, "failed to connect to destination "
-				"pool: %d\n", rc);
-			D_GOTO(out, rc);
+		if (!uuid_is_null(ca->dst_p_uuid)) {
+			rc = daos_pool_connect(ca->dst_p_uuid, sysname,
+					       DAOS_PC_RW, &ca->dst_poh, NULL,
+					       NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to connect to "
+					"destination pool: %d\n", rc);
+				D_GOTO(out, rc);
+			}
+		} else {
+			uuid_copy(ca->dst_p_uuid, ca->src_p_uuid);
+			rc = daos_pool_connect(ca->dst_p_uuid, sysname,
+					       DAOS_PC_RW, &ca->dst_poh,
+					       NULL, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to connect to "
+					"destination pool: %d\n", rc);
+				D_GOTO(err_src, rc);
+			}
+			uuid_copy(dattr.da_puuid, ca->dst_p_uuid);
+			uuid_copy(dattr.da_cuuid, ca->dst_c_uuid);
+			dattr.da_type = ca->cont_layout;
+			dattr.da_props = props;
+			rc = duns_create_path(ca->dst_poh, path,
+					      &dattr);
+			if (rc != 0) {
+				fprintf(stderr, "please provide a destination "
+					"pool or UNS path of the form:\n\t"
+					"--dst </$pool> | </path/to/uns>\n");
+					D_GOTO(err_dst_root, rc);
+			}
+			uuid_copy(ca->dst_c_uuid, dattr.da_cuuid);
 		}
-		rc = daos_cont_open(fa->dst_poh, fa->dst_c_uuid, DAOS_COO_RW,
-				    &fa->dst_coh, dst_cont_info, NULL);
+		rc = daos_cont_open(ca->dst_poh, ca->dst_c_uuid,
+				    DAOS_COO_RW, &ca->dst_coh,
+				    dst_cont_info, NULL);
 		if (rc != 0) {
```

Comment:
TBH i'm not understanding what is going on here.
you need some comments in the code to explain the logic.
why do we attempt to create on every error?

Reply:
@mchaarawi the first check is if the dst pool uuid is null, and if it is then this is a UNS destination path, and I have to copy the src pool uuid into the dst pool uuid, then create the UNS path on the destination. I first try a daos_cont_open because it could be a filesystem copy, and it is not required that the dest container doesn't already exist, and if it fails then I need to create the destination container (whether it is for the filesystem or clone copy). I then have to try to open the container again after it is created. I am adding some comments to hopefully make this more clear.

Reply:
the issue is that you are issuing a create for **any** type of error, not just DER_NOEXIST errors!
and yes some comments are needed here to help understand what is going on.

Reply:
@mchaarawi ah yeah I somehow missed that! I will update that, thanks.